### PR TITLE
feat(servicestage): add new environment resource support

### LINKED ISSUE #383: No resources related to ServiceStage
None of the ServiceStage related resources are available.

### Terraform Version
```
terraform version
Terraform v0.12.28
+ provider.huaweicloud v1.16.0
```

### References

* [ServiceStage elements in AOS (Application Orchestration Service)](https://support.huaweicloud.com/en-us/tr-aos/aos_01_4013.html)
  * ServiceStage.Agent
  * ServiceStage.AppGroup
  * ServiceStage.ContainerComponent
  * ServiceStage.Job
  * ServiceStage.StatefulApplication
  * ServiceStage.StatelessApplication
* [Application Management V2 APIs](https://support.huaweicloud.com/intl/en-us/api-servicestage/servicestage_api_0091.html)

### PR DESCRIPTION
**Pull request by Lance52259**
**What this PR does / why we need it**:
**Background:**
Environment is a collection of infrastructures, covering computing, storage, and networks, used for application deployment and running.
ServiceStage combines basic resources, such as Cloud Container Engine (CCE) and Elastic Cloud Server (ECS) and optional resources, such as Elastic Load Balance (ELB), Relational Database Service (RDS), and Distributed Cache Service (DCS) in the same VPC into an environment, such as the development environment, testing environment, pre-production environment, and production environment.
Networks in an environment can communicate with each other.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #383 
reference #2030 

**Special notes for your reviewer**:

**Release note**:

```release-note
1. add new environment resource support.
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud/services/acceptance/servicestage' TESTARGS='-run=TestAccServiceStageEnvironment_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/servicestage -v -run=TestAccServiceStageEnvironment_basic -timeout 360m -parallel 4
=== RUN   TestAccServiceStageEnvironment_basic
=== PAUSE TestAccServiceStageEnvironment_basic
=== CONT  TestAccServiceStageEnvironment_basic
--- PASS: TestAccServiceStageEnvironment_basic (683.65s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/servicestage      683.775s
```

### CODE REVIEW

File: huaweicloud/config/endpoints.go [hunk 1]
```diff
@@ -423,6 +423,10 @@ var allServiceCatalog = map[string]ServiceCatalog{
 		Name:    "dms",
 		Version: "v2",
 	},
+	"servicestagev2": {
```
Comment (hmbig2) [inline huaweicloud/config/endpoints.go:513]: just one endpoint , maybe can remove version

File: huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+	return environments.CreateOpts{
+		Name:              d.Get("name").(string),
+		Description:       &desc,
+		VpcId:             d.Get("vpc_id").(string),
+		BaseResources:     buildResourcesList(d.Get("basic_resources").([]interface{})),
+		OptionalResources: buildResourcesList(d.Get("optional_resources").([]interface{})),
+	}
+}
+
+func resourceServiceStageEnvCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
```
Comment (hmbig2) [inline huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go]: remove package name

File: huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+	config := meta.(*config.Config)
+	client, err := config.ServiceStageV2Client(config.GetRegion(d))
+	if err != nil {
+		return fmtp.DiagErrorf("error creating ServiceStage v2 client: %s", err)
+	}
+
+	opt := buildServiceStageEnvCreateOpts(d)
+	logp.Printf("[DEBUG] The createOpt of ServiceStage environment is: %s", opt)
```
Comment (hmbig2) [inline huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go]: s/logp.Printf/log.printf

Comment (Lance52259) [reply huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go]: yeah, and the type of arg opt is wrong.

File: huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+	// Currently, the base resource type supports CCE, CCI, ECS and AS.
+	bTypes := []string{"cce", "cci", "ecs", "as"}
+	// Currently, the optional resource type supports ELB, EIP, RDS, DCS and CSE.
+	oTypes := []string{"elb", "eip", "rds", "dcs", "cse"}
+	mErr := multierror.Append(nil,
+		d.Set("region", region),
+		d.Set("name", resp.Name),
+		d.Set("description", resp.Description),
+		d.Set("vpc_id", resp.VpcId),
+		d.Set("basic_resources", flattenEnvResources(resp.BaseResources, bTypes)),
```
Comment (hmbig2) [inline huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go]: Duplicated with line 230 ? basic_resources  and optional_resources,

Comment (Lance52259) [reply huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go]: yeah, the codes of line 226 to 227 are useless.

File: huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go (hunk 2 above)
Comment (Lance52259) [reply huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go]: done.

**Comment by Lance52259** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/servicestage' TESTARGS='-run=TestAccServiceStageEnvironment_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/servicestage -v -run=TestAccServiceStageEnvironment_basic -timeout 360m -parallel 4
=== RUN   TestAccServiceStageEnvironment_basic
=== PAUSE TestAccServiceStageEnvironment_basic
=== CONT  TestAccServiceStageEnvironment_basic
--- PASS: TestAccServiceStageEnvironment_basic (600.49s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/servicestage      601.036s
```

File: huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+		rmResult[i] = environments.Resource{
+			ID:   v.(string),
+			Type: rType,
+		}
+	}
+
+	return addResult, rmResult
+}
+
+func getBaseResourcesChange(d *schema.ResourceData) (baseRes, rmRes []environments.Resource) {
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go]: suggest to use bTypes here

File: huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+	if d.HasChange("basic_resources.0.cce_ids") {
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go]: how about use a for loop

File: huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+	if d.HasChange("basic_resources.0.as_ids") {
+		oldSet, newSet := d.GetChange("basic_resources.0.as_ids")
+		base, rm := getResourcesChange(oldSet, newSet, "as")
+		baseRes = append(baseRes, base...)
+		rmRes = append(rmRes, rm...)
+	}
+	return
+}
+
+func getOptionalResourcesChange(d *schema.ResourceData) (optRes, rmRes []environments.Resource) {
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go]: please consider merge this func with getBaseResourcesChange

File: huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+
+	resp, err := environments.Get(client, d.Id())
+	if err != nil {
+		common.CheckDeletedDiag(d, err, "error retrieving ServiceStage environment")
+	}
+
+	// Currently, the base resource type supports CCE, CCI, ECS and AS.
+	bTypes := []string{"cce", "cci", "ecs", "as"}
+	// Currently, the optional resource type supports ELB, EIP, RDS, DCS and CSE.
+	oTypes := []string{"elb", "eip", "rds", "dcs", "cse"}
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go]: make bTypes and oTypes consist so they can be used in other funcs

Comment (Lance52259) [reply huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go]: done.

File: huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go (hunk 7 above)
Comment (Lance52259) [reply huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go]: done.

File: huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go (hunk 6 above)
Comment (Lance52259) [reply huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go]: done.

File: huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go (hunk 5 above)
Comment (Lance52259) [reply huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go]: This function has been refactored.

File: docs/resources/servicestage_environment.md [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+## Example Usage
+
+### Create an environment based on some cce clusters
+
+```hcl
+variable "env_name" {}
+variable "vpc_id" {}
+
+resource "huaweicloud_cce_cluster" "test" {
+  ...
```
Comment (niuzhenguo) [inline docs/resources/servicestage_environment.md]: suggest to also define cce cluster as a variable like vpc_id or add all parameters to the script.

File: docs/resources/servicestage_environment.md [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `region` - (Optional, String, ForceNew) Specifies the region in which to create the ServiceStage environment.
+  If omitted, the provider-level region will be used. Changing this will create a new resource.
+
+* `name` - (Required, String) Specifies the environment name.
+  The name can contains of 2 to 64 characters, only letters, digits, hyphens (-) and underscores (_) are allowed.
```
Comment (niuzhenguo) [inline docs/resources/servicestage_environment.md]: s/contains/contain

File: huaweicloud/config/endpoints.go (hunk 1 above)
Comment (Lance52259) [reply huaweicloud/config/endpoints.go:513]: There will also be a v1 endpoint in the future, I want to use it as default.

**Comment by Lance52259** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/servicestage' TESTARGS='-run=TestAccEnvironment_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/servicestage -v -run=TestAccEnvironment_basic -timeout 360m -parallel 4
=== RUN   TestAccEnvironment_basic
=== PAUSE TestAccEnvironment_basic
=== CONT  TestAccEnvironment_basic
--- PASS: TestAccEnvironment_basic (1222.26s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/servicestage        1223.425s
```

File: docs/resources/servicestage_environment.md (hunk 9 above)
Comment (Lance52259) [reply docs/resources/servicestage_environment.md]: done.

File: docs/resources/servicestage_environment.md (hunk 10 above)
Comment (Lance52259) [reply docs/resources/servicestage_environment.md]: done.

File: docs/resources/servicestage_environment.md [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+
+* `description` - (Optional, String) Specifies the environment description.
+  The description can contian a maximum of 128 characters.
+
+<a name="servicestage_env_basic_resources"></a>
+The `basic_resources` block supports:
+
+* `cce_ids` - (Optional, List) Specifies the ID list of CCE clusters under the same VPC as the environment.
+
+* `cci_ids` - (Optional, List) Specifies the **name** list of CCI namespaces under the same VPC as the environment.
```
Comment (niuzhenguo) [inline docs/resources/servicestage_environment.md]: how about just naming these parameters without `_ids`, seems cce = [aa, bb], cci = [cc, dd] more concise.

File: huaweicloud/config/endpoints.go [hunk 12]
```diff
@@ -16,27 +16,28 @@ type ServiceCatalog struct {
...
+	"waf":          {"waf-dedicated"},
+	"geminidb":     {"geminidbv31"},
+	"dli":          {"dliv2"},
+	"dcs":          {"dcsv1"},
+	"dis":          {"disv3"},
+	"dms":          {"dmsv2"},
+	"dws":          {"dwsv2"},
+	"apig":         {"apigv2"},
+	"modelarts":    {"modelartsv2"},
+	"servicestage": {"servicestagev2"},
```
Comment (niuzhenguo) [inline huaweicloud/config/endpoints.go:40]: why we have multi catalog here? seems we can just use one.

File: docs/resources/servicestage_environment.md [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+  The [object](#servicestage_env_basic_resources) structure is documented below.
+
+* `optional_resources` - (Optional, List) Specifies the optional resources.
+  The [object](#servicestage_env_optional_resources) structure is documented below.
+
+* `description` - (Optional, String) Specifies the environment description.
+  The description can contian a maximum of 128 characters.
+
+<a name="servicestage_env_basic_resources"></a>
+The `basic_resources` block supports:
```
Comment (niuzhenguo) [inline docs/resources/servicestage_environment.md]: I would like to just keep the API block here.
```
"base_resources": [
		{
			"id": "523498f1-36c4-11eb-ae36-0255ac1000c2",
			"type": "cce",
			"name": "cce-test"
		}
	]
```

File: huaweicloud/config/endpoints.go (hunk 12 above)
Comment (Lance52259) [reply huaweicloud/config/endpoints.go:40]: the authorization resources use ver.1 API

File: docs/resources/servicestage_environment.md (hunk 13 above)
Comment (Lance52259) [reply docs/resources/servicestage_environment.md]: done.

**Comment by Lance52259** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/servicestage' TESTARGS='-run=TestAccEnvironment_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/servicestage -v -run=TestAccEnvironment_basic -timeout 360m -parallel 4
=== RUN   TestAccEnvironment_basic
=== PAUSE TestAccEnvironment_basic
=== CONT  TestAccEnvironment_basic
--- PASS: TestAccEnvironment_basic (943.40s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/servicestage943.462s
```

File: docs/resources/servicestage_environment.md [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+  The [object](#servicestage_env_resources) structure is documented below.
+
+* `optional_resources` - (Optional, List) Specifies the optional resources.
+  The [object](#servicestage_env_resources) structure is documented below.
+
+* `description` - (Optional, String) Specifies the environment description.
+  The description can contian a maximum of 128 characters.
+
+<a name="servicestage_env_resources"></a>
+The `basic_resources` and optional_resources` block supports:
```
Comment (niuzhenguo) [inline docs/resources/servicestage_environment.md]: missing a left ` on optional_resources

Comment (Lance52259) [reply docs/resources/servicestage_environment.md]: yeah

File: huaweicloud/services/acceptance/servicestage/resource_huaweicloud_servicestage_environment_test.go [hunk 15]
```diff
@@ -0,0 +1,582 @@
...
+    id   = huaweicloud_cci_namespace.test[0].name
+  }
+  basic_resources {
+    type = "ecs"
+    id   = huaweicloud_compute_instance.test[0].id
+  }
+  basic_resources {
+    type = "cce"
+    id   = huaweicloud_as_group.test[0].id
+  }
```
Comment (ShiChangkuo) [inline huaweicloud/services/acceptance/servicestage/resource_huaweicloud_servicestage_environment_test.go:492]: the type and id do not match, please check it

File: huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go [hunk 16]
```diff
@@ -0,0 +1,241 @@
...
+					validation.StringMatch(regexp.MustCompile(`^[A-Za-z]([\w-]*[A-Za-z0-9])?$`),
+						"The name must start with a letter and end with a letter or digit, and can only contain "+
+							"letters, digits, underscores (_) and hyphens (-)."),
+					validation.StringLenBetween(2, 64),
+				),
+			},
+			"description": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.StringLenBetween(1, 128),
```
Comment (ShiChangkuo) [inline huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go]: we should allow to clear the `description` with ```description = ""```

File: huaweicloud/services/acceptance/servicestage/resource_huaweicloud_servicestage_environment_test.go (hunk 15 above)
Comment (Lance52259) [reply huaweicloud/services/acceptance/servicestage/resource_huaweicloud_servicestage_environment_test.go:492]: fixed.

File: huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go (hunk 16 above)
Comment (Lance52259) [reply huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go]: fixed.

**Comment by Lance52259** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/servicestage' TESTARGS='-run=TestAccEnvironment_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/servicestage -v -run=TestAccEnvironment_basic -timeout 360m -parallel 4
=== RUN   TestAccEnvironment_basic
=== PAUSE TestAccEnvironment_basic
=== CONT  TestAccEnvironment_basic
--- PASS: TestAccEnvironment_basic (1144.81s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/servicestage      1144.959s
```

**Pull request merged**